File: 17/computer.py
import sys
import time
import re
import timeit

def _run(a,b,c, program, check=False, seen=None):
    output = []

    ip = 0

    while ip < len(program):

        # x = str((a,b,c,ip,output))
        # if x in seen:
        #     # print("seen")
        #     return {}, None
        # seen.add(x)

        opcode = program[ip]

        if opcode in [0,2,5,6,7]:
            combo = program[ip+1]
            if combo == 4:
                combo = a
            elif combo == 5:
                combo = b
            elif combo == 6:
                combo = c

        # o = []
        # 2,4 => b = a % 8
        # 1,1 => b = b & 1
        # 7,5 => c = a // 2**b
        # 0,3 => a = a // 2**3
        # 1,4 => b = b & 4
        # 4,5 => b = b & c
        # 5,5 => o <- b % 8
        # 3,0 => a != 0 GOTO a

        match opcode:
            case 0: #adv
                a = a // (1 << combo)
            case 1:
                b ^= program[ip+1]
            case 2: #bst
                b = (combo % 8) & 1
                ip += 4
                continue
            case 3: #jnz
                if not a == 0:
                    ip = program[ip+1]
                    continue
            case 4:
                b ^= c
            case 5:
                if check and combo % 8 != program[len(output)]:
                    return {}, output

                output.append(combo % 8)
                # if check:
                #     if len(output) > len(program):
                #         return {}, output
                #     for i in range(len(output)):
                #         if output[i] != program[i]:
                #             return {}, output

            case 6: #bdv
                b = a // (1 << combo)

            case 7: #cdv
                c = a // (1 << combo)

        ip += 2

    return {'A':a, 'B': b, 'C':c}, output

def run(input):
    registers = {}
    program = []

    for line in re.findall("(.+?) ?([A-Z])?: (.+)", input):
        if line[0] == 'Register':
            registers[line[1]] = int(line[2])
        if line[0] == 'Program':
            program += [int(x) for x in line[2].split(",")]

    seen = set()
    registers, output = _run(*registers.values(), program, False, seen)

    return registers, output

def find(input, start, limit):
    registers = {}
    program = []

    for line in re.findall("(.+?) ?([A-Z])?: (.+)", input):
        if line[0] == 'Register':
            registers[line[1]] = int(line[2])
        if line[0] == 'Program':
            program += [int(x) for x in line[2].split(",")]

    i = 0
    try:
        seen = set()
        for i in range(start, limit, 8):
            registers, output = _run(i, 0, 0, program, True, seen)
            if output == program:
                print('answer', i)
                break
        else:
            print('fail')
    except:
        print()
        print(i)
        open("state",'w+').write(str(i-1))
        print()
        raise



# registers, output = run("""
# Register A: 729
# Register B: 0
# Register C: 0

# Program: 0,1,5,4,3,0""")
# print(output)
# assert output == [int(x) for x in "4,6,3,5,6,3,5,2,1,0".split(',')]

# registers, output = run("""
# Register A: 0
# Register B: 0
# Register C: 9

# Program: 2,6""")
# print(registers)
# assert registers['B'] == 1

# registers, output = run("""
# Register A: 10
# Register B: 0
# Register C: 0

# Program: 5,0,5,1,5,4""")
# print(output)
# assert output == [int(x) for x in "0,1,2".split(',')]

# registers, output = run("""
# Register A: 2024
# Register B: 0
# Register C: 0

# Program: 0,1,5,4,3,0""")
# print(output)
# assert output == [int(x) for x in "4,2,5,6,7,7,7,7,3,1,0".split(',')]

# registers, output = run("""
# Register A: 0
# Register B: 29
# Register C: 0

# Program: 1,7""")
# print(registers)
# assert registers['B'] == 26

# IN = open("input.txt").read()
# registers, output = run(IN)
# print(output)
# assert output == [int(x) for x in "4,0,4,7,1,2,7,1,6".split(',')]

# registers, output = run("""
# Register A: 2024
# Register B: 0
# Register C: 0

# Program: 0,3,5,4,3,0""")
# print(output)
# assert output != [int(x) for x in "0,3,5,4,3,0".split(',')]

# registers, output = run("""
# Register A: 117440
# Register B: 0
# Register C: 0

# Program: 0,3,5,4,3,0""")
# print(output)
# assert output == [int(x) for x in "0,3,5,4,3,0".split(',')]



speed = 0
for i in range(10):
    start = time.time()
    find("""
Register A: 2024
Register B: 0
Register C: 0

Program: 0,3,5,4,3,0""", 0, 117441)
    stop = time.time()
    speed += (stop - start) / 10
print("speed", speed)

start = time.time()
i = int(open("state",'r').read())
registers, output = run(open("input.txt").read())
print(registers, output)
assert registers == {'A': 0, 'B': 6, 'C': 0}
assert output ==  [4, 0, 4, 7, 1, 2, 7, 1, 6]

print(i)
print()
print()
print()
find(open("input.txt").read(), i, int(sys.maxsize))
stop = time.time()
print("speed", stop - start)

File: 17/computer.go
package main

import (
	"fmt"
	"time"
	// "sync"
)

func run(a uint64, b uint64, c uint64, program []uint8, check []uint8 ) bool {
	ip := uint8(0)
	output := []uint8{}

	if len(check) == 0 {
		check = program
	}

	for ip < uint8(len(program)) {
		// fmt.Printf("%20d %20d %20d %20.d %v\n", a,b,c,ip,output)
		opcode := program[ip]
		combo := uint64(program[ip+1])

		if combo == 4 {
			combo = a
		} else if combo == 5 {
			combo = b
		} else if combo == 6 {
			combo = c
		}

		switch opcode {
			case 0:
				a = a / (1 <<combo)
			case 1:
				b = b ^ uint64(program[ip+1])
			case 2:
				b = uint64(combo % 8)
			case 3:
				if a != 0 {
					ip = program[ip+1]
					continue
				}
			case 4:
				b = b ^ c
			case 5:
				if check[len(output)] != uint8(combo % 8) {
					return false
				}
				output = append(output, uint8(combo % 8))
				// fmt.Println()
				// fmt.Printf("%20d %20d %20d %20.d %v\n", a,b,c,ip,output)

				if len(output) == len(check){
					return true
				}
			case 6:
				b = a / (1<<combo)
			case 7:
				c = a / (1<<combo)
		}

		ip += 2
	}

	return false
}

// func main(){
// 	start := time.Now()
// 	for i := uint64(0); i <= 18446744073709551615; i+=8 {
// 		if run(i, 0, 0, []uint8{0,3,5,4,3,0}){
// 			fmt.Println(i)
// 			break
// 		}
// 	}
// 	fmt.Println(time.Since(start))

// 	for i := uint64(4294967295); i <= 18446744073709551615; i+=8 {
// 		if run(i, 0, 0, []uint8{2,4,1,1,7,5,0,3,1,4,4,5,5,5,3,0}){
// 			fmt.Println(i)
// 			break
// 		}
// 	}
// }

// func loop(start uint64, end uint64){
// 	for i := uint64(start); i <= end; i++ {
// 		if run(i, 0, 0, []uint64{2,4,1,1,7,5,0,3,1,4,4,5,5,5,3,0}){
// 			fmt.Println(i)
// 			break
// 		}
// 	}
// }

func main(){
	start := time.Now()
	for i := uint64(0); i <= 18446744073709551615; i += 64 {
		if run(i, 0, 0, []uint8{0,3,5,4,3,0}, []uint8{})  {
			fmt.Println(i)
			break
		}
	}
	fmt.Println(time.Since(start))


	{
		start := time.Now()
		if ! run(51571418, 0, 0, []uint8{2,4,1,1,7,5,0,3,1,4,4,5,5,5,3,0}, []uint8{4,0,4,7,1,2,7,1,6}) {
			panic(1)
		}
		fmt.Println(time.Since(start))
	}
	{
		start := time.Now()
		if ! run(202322348616234, 0, 0, []uint8{2,4,1,1,7,5,0,3,1,4,4,5,5,5,3,0}, []uint8{}) {
			panic(1)
		}
		fmt.Println(time.Since(start))
	}

	program := []uint8{2,4,1,1,7,5,0,3,1,4,4,5,5,5,3,0}
	x := uint64(0)
	for j := len(program) - 1; j > 0; j-- {
		for i := uint64(x); i < 1<<63; i++  {
			if run(i, 0, 0, program, program[j:]) {
				fmt.Println(i)
				x = i*uint64(len(program)/2)
				break
			}
		}
	}
	{
		program := []uint8{0,3,5,4,3,0}
		x := uint64(0)
		for j := len(program); j > 0; j-- {
			for i := uint64(x); i < 1<<63; i++  {
				if run(i, 0, 0, program, program[j-1:]) {
					fmt.Println(i)
					x = i*uint64(len(program)/2)
					break
				}
			}
		}

	}

	// {
	// 	for i := uint64(40); i < 1<<63; i++  {
	// 		if run(i, 0, 0, []uint8{2,4,1,1,7,5,0,3,1,4,4,5,5,5,3,0}, []uint8{3,0}) {
	// 			fmt.Println(i)
	// 			break
	// 		}
	// 	}
	// }

	// {
	// 	for i := uint64(46); i < 1<<63; i++  {
	// 		if run(i, 0, 0, []uint8{2,4,1,1,7,5,0,3,1,4,4,5,5,5,3,0}, []uint8{5,3,0}) {
	// 			fmt.Println(i)
	// 			break
	// 		}
	// 	}
	// }

	// {
	// start := time.Now()
	// for i := uint64(0); i <= 18446744073709551615; i += 117440 {
	// 	if run(i, 0, 0, []uint8{0,3,5,4,3,0}){
	// 		fmt.Println(i)
	// 	}
	// }
	// fmt.Println(time.Since(start))
	// }
	// {
	// 	ch := make(chan uint64)

	// 	start := time.Now()
	// 	for i := uint64(0); i <= 117440*2; i = i + 117440*2 / 4 {

	// 		go func(start uint64, end uint64){
	// 			for i := start; i < end; i=i+8 {
	// 				if run(i, 0, 0, []uint8{0,3,5,4,3,0}){
	// 					// fmt.Println(i)
	// 					ch <- i
	// 					// close(ch)
	// 				}
	// 			}

 // 			}(i, i + 117440*2 / 4)
	// 	}
	// 	for i := range ch {
	// 		fmt.Println(i)
	// 		close(ch)
	// 	}
	// 	fmt.Println(time.Since(start))
	// }

	// {
	// 	ch := make(chan uint64)
	// 	for i := uint64(44); i < 50; i++  {
	// 		start := uint64(1<<i)
	// 		end := uint64(1<<(i+1))

	// 		go func(start uint64, end uint64){
	// 			fmt.Println(start, end)
	// 			for i := start; i < end; i=i+8 {
	// 				if run(i, 0, 0, []uint8{2,4,1,1,7,5,0,3,1,4,4,5,5,5,3,0}){
	// 					fmt.Println(i)
	// 					ch <- i
	// 				}
	// 			}
	// 			ch <- 0
 // 			}(start, end)
	// 	}
	// 	for i := range ch {
	// 		fmt.Println(i)
	// 	}
	// }

	// {
	// 	var wg sync.WaitGroup
	// 	ch := make(chan uint64)
	// 	max := uint64(1<<63)
	// 	for i := max; i > 0; i = i / 2  {
	// 		fmt.Println(i, max/i)
	// 		// fmt.Println()
	// 		for j := max; j > 0; j = j + i {
	// 			start := j+i
	// 			end := j
	// 			step := i
	// 			// fmt.Println(start,end, step)

	// 			wg.Add(1)
	// 			go func(start uint64, end uint64){
	// 				defer wg.Done()
	// 				for x := start ; x < end; x+=step*2 {
	// 					if run(x, 0, 0, []uint8{2,4,1,1,7,5,0,3,1,4,4,5,5,5,3,0}, []uint8{}){
	// 						fmt.Println(x)
	// 						ch <- x
	// 					}
	// 				}
	//  			}(start, end)

	// 		}
	// 	}
	// 		// for j := uint64(0); j < 1<<62; j+=i {
	// 		// fmt.Println(j)
	// 		// start := uint64(j<<i)
	// 		// end := uint64(j<<(i+1))
	// 		// fmt.Println(start, end)
	// 		// // fmt.Println(i)
	// 	go func() { wg.Wait(); close(ch)}()
	// 	for i := range ch {
	// 		fmt.Println(i)
	// 	}
	// }
}
